cli/internal/fs/turbo_json.go

```diff
@@ -248,82 +293,169 @@ func (pc Pipeline) HasTask(task string) bool {
+
+		if bookkeepingTaskDef.hasField("TaskDependencies") {
+			mergedTaskDefinition.TaskDependencies = taskDef.TaskDependencies
+		}
+
+		if bookkeepingTaskDef.hasField("Inputs") {
+			mergedTaskDefinition.Inputs = taskDef.Inputs
+		}
+
+		if bookkeepingTaskDef.hasField("OutputMode") {
+			mergedTaskDefinition.OutputMode = taskDef.OutputMode
+		}
+		mergedTaskDefinition.Persistent = taskDef.Persistent
+	}
+
+	return mergedTaskDefinition, nil
+}
+
+// UnmarshalJSON deserializes a single task definition from
+// turbo.json into a TaskDefinition struct
+func (btd *BookkeepingTaskDefinition) UnmarshalJSON(data []byte) error {
 	task := rawTask{}
 	if err := json.Unmarshal(data, &task); err != nil {
 		return err
 	}
 
-	var inclusions []string
-	var exclusions []string
+	btd.fieldsMeta = map[string]bool{}
 
 	if task.Outputs != nil {
+		var inclusions []string
+		var exclusions []string
+		// Assign a bookkeeping field so we know that there really were
+		// outputs configured in the underlying config file.
+		btd.fieldsMeta["Outputs"] = true
+
 		for _, glob := range task.Outputs {
 			if strings.HasPrefix(glob, "!") {
 				exclusions = append(exclusions, glob[1:])
 			} else {
 				inclusions = append(inclusions, glob)
 			}
 		}
-	}
 
-	c.Outputs = TaskOutputs{
-		Inclusions: inclusions,
-		Exclusions: exclusions,
+		btd.TaskDefinition.Outputs = TaskOutputs{
+			Inclusions: inclusions,
+			Exclusions: exclusions,
+		}
+
+		sort.Strings(btd.TaskDefinition.Outputs.Inclusions)
+		sort.Strings(btd.TaskDefinition.Outputs.Exclusions)
 	}
 
-	sort.Strings(c.Outputs.Inclusions)
-	sort.Strings(c.Outputs.Exclusions)
 	if task.Cache == nil {
-		c.ShouldCache = true
+		btd.TaskDefinition.ShouldCache = true
 	} else {
-		c.ShouldCache = *task.Cache
+		btd.TaskDefinition.ShouldCache = *task.Cache
 	}
 
 	envVarDependencies := make(util.Set)
 
-	c.TopologicalDependencies = []string{} // TODO @mehulkar: this should be a set
-	c.TaskDependencies = []string{}        // TODO @mehulkar: this should be a set
+	btd.TaskDefinition.TopologicalDependencies = []string{} // TODO @mehulkar: this should be a set
+	btd.TaskDefinition.TaskDependencies = []string{}        // TODO @mehulkar: this should be a set
 
 	for _, dependency := range task.DependsOn {
 		if strings.HasPrefix(dependency, envPipelineDelimiter) {
 			log.Printf("[DEPRECATED] Declaring an environment variable in \"dependsOn\" is deprecated, found %s. Use the \"env\" key or use `npx @turbo/codemod migrate-env-var-dependencies`.\n", dependency)
 			envVarDependencies.Add(strings.TrimPrefix(dependency, envPipelineDelimiter))
 		} else if strings.HasPrefix(dependency, topologicalPipelineDelimiter) {
-			c.TopologicalDependencies = append(c.TopologicalDependencies, strings.TrimPrefix(dependency, topologicalPipelineDelimiter))
+			// Assign bookkeeping, but only once, since we are in a loop
+			if _, ok := btd.fieldsMeta["TopologicalDependencies"]; !ok {
+				btd.fieldsMeta["TopologicalDependencies"] = true
+			}
```

Comment:
I'm not sure I understand why we are careful to only assign this once since we're doing a hash lookup already in every iteration. Minor taste suggestion to cut down on the code nesting.
```suggestion	
			btd.fieldsMeta["TopologicalDependencies"] = true

```

Reply:
i thought it would be better in case the logic changed to also assign `false` in some case, so that we don't have a last-one-wins scenario.  Also just felt more "correct" to assign once since we only care about the first one we encounter. you're right that it doesn't matter right now though, I've taken your suggestion!